### PR TITLE
Streamline embeddings from "non-embedding" models

### PR DESCRIPTION
**Pull request by iamlemec**
The goal here is to get the big embedding models at the top of the MTEB leaderboard working. There are two changes:

- Fix an inconsistency in output counting for u_batches. Now `batch.logits` is fully ignored for pooled embeddings.
- Add an `attention_type` to `llama_contex_params` that allows for causal, non-causal, or unspecified (model default).

With this PR, we can get accurate results (matching HF) from at least the number 2 spot `gte-Qwen2-7B-instruct`. For instance, with the command:
```bash
./llama-embedding -m gte-qwen2-7b-instruct-f16.gguf -p "hello world" -ngl 99 --pooling last --at
tention non-causal -c 512
```

<br/>

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [X] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -12618,14 +12618,15 @@ static int llama_decode_internal(
     std::vector<llama_seq_id *>            seq_id_arr;
     std::vector<std::vector<llama_seq_id>> seq_id;
 
+    // this indicates we are doing pooled embedding, so we ignore batch.logits and output all tokens
+    bool embed_pooled = cparams.embeddings && cparams.pooling_type != LLAMA_POOLING_TYPE_NONE;
```
Comment (ggerganov) [inline llama.cpp:12622]: Naming this `embd_pooled` seems more inline with the usage of the `embd` identifier in `llama.cpp`:

```suggestion
    const bool embd_pooled = cparams.embeddings && cparams.pooling_type != LLAMA_POOLING_TYPE_NONE;
```

Comment (iamlemec) [reply llama.cpp]: yup, agreed.

File: common/common.cpp [hunk 2]
```diff
@@ -472,6 +472,17 @@ bool gpt_params_find_arg(int argc, char ** argv, const std::string & arg, gpt_pa
         else { invalid_param = true; }
         return true;
     }
+    if (arg == "--attention") {
+        if (++i >= argc) {
+            invalid_param = true;
+            return true;
+        }
```
Comment (compilade) [inline common/common.cpp:479]: This might be able to use `CHECK_ARG`

**Review by compilade — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]

**Review by compilade — APPROVED** [review]
Still looks good to me.

@iamlemec, you'll need to merge `master` into to this first to resolve the conflicts; some files have moved.

**Comment by iamlemec** [comment]
@compilade cool! just rebased to master
